File: symbolRecognizer2/fv.c
/*****
fv.c	Creates a feature vector, useful for gesture classification,
from a sequence of points (e.g. mouse points).
****/

#include <stdio.h>
#include <math.h>
#include "matrix.h" /* contains Vector and associated functions */
#include "fv.h"

/* runtime settable parameters */
double dist_sq_threshold = DIST_SQ_THRESHOLD;
double se_th_rolloff = SE_TH_ROLLOFF;

#define EPS (1.0e-4)

/* allocate an FV struct including feature vector */
FV FvAlloc()
{
    register FV fv = (FV) mallocOrDie(sizeof(struct fv));
    fv->y = NewVector(NFEATURES);
    FvInit(fv);
    return fv;
}

/* free memory associated with an FV struct */
void
FvFree(fv)
FV fv;
{
    FreeVector(fv->y);
    free((char *) fv);
}

/* initialize an FV struct to prepare for incoming gesture points */
void
FvInit(fv)
register FV fv;
{
    register int i;

    fv->npoints = 0;
    fv->initial_sin = fv->initial_cos = 0.0;
    fv->maxv = 0;
    fv->path_r = 0;
    fv->path_th = 0;
    fv->abs_th = 0;
    fv->sharpness = 0;
    fv->maxv = 0;
    for(i = 0; i < NFEATURES; i++)
        fv->y[i] = 0.0;
}

/* update an FV struct to reflect a new input point */
void
FvAddPoint(fv, x, y, t)
register FV fv; int x, y; long t;
{
    double dx1, dy1, magsq1;

    double th, absth, d;
#ifdef PF_MAXV
    long lasttime;
#endif

    ++fv >npoints;
    if(fv->npoints == 1) {/* first point, initialize some vars  */
        fv->starttime = fv->endtime = t;
        fv->startx = fv->endx = fv->minx = fv->maxx = x;
        fv->starty = fv->endy = fv->miny = fv->maxy = y;
        fv->endx = x;
        fv->endy = y;
        return;
    }
    dx1 = x-fv->endx;
    dy1 = y-fv->endy;
    magsq1 = dx1*dx1 + dy1*dy1;

    if(magsq1 <= dist_sq_threshold){
        fv->npoints--;
        return;	/* ignore a point close to the last point */
    }
    if(x < fv->minx) fv->minx = x;
    if(x > fv->maxx) fv->maxx = x;
    if(y < fv->miny) fv->miny = y;
    if(y > fv->maxy) fv->maxy = y;
#ifdef PF_MAXV
    lasttime = fv->endtime;
#endif
    fv->endtime = t;

    d = sqrt(magsq1);
    fv->path_r += d;	/* update path length feature */

    /* calculate initial theta when the third point is seen */
    if(fv->npoints == 3){
        double magsq, dx, dy, recip;
        dx = x-fv->startx; dy = y-fv->starty;
        magsq = dx*dx + dy*dy;
        if(magsq > dist_sq_threshold){
            /* find angle w.r.t. positive x axis e.g. (1,0) */
            recip = 1 / sqrt(magsq);
            fv->initial_cos = dx*recip;
            fv->initial_sin = dy*recip;
        }
    }
    if(fv->npoints >= 3){
        /* update angle based features */
        th = absth = atan2(dx1*fv->dy2-fv->dx2*dy1, dx1*fv->dx2 + dy1*fv->dy2);
        if(absth < 0) absth = -absth;
        fv->path_th += th;
        fv->abs_th += absth;
        fv->sharpness += th*th;

#ifdef PF_MAXV /* compute max velocity */
        if(fv->endtime > lasttime &&  (v = d / (fv->endtime-lasttime)) > fv->maxv)
            fv->maxv = v;
#endif
    }
    /* prepare for next iteration */
    fv->endx = x; fv->endy = y;
    fv->dx2 = dx1; fv->dy2 = dy1;
    fv->magsq2 = magsq1;

    return;
}

/* calculate and return a feature vector */
Vector
FvCalc(fv)
register FV fv;
{
    double bblen, selen, factor;
    if(fv->npoints <= 1)
        return fv->y;	/* a feature vector of all zeros */
    fv->y[PF_INIT_COS] = fv->initial_cos;
    fv->y[PF_INIT_SIN] = fv->initial_sin;
    /* compute the length of the bounding box diagonal */
    bblen = hypot(fv->maxx-fv->minx, fv->maxy-fv->miny);

    fv->y[PF_BB_LEN] = bblen;

    /* the bounding box angle defaults to 0 for small gestures */
    if(bblen*bblen > dist_sq_threshold)
        fv->y[PF_BB_TH] = atan2(fv->maxy-fv->miny,
                                fv->maxx-fv->minx);

    /* compute the length and angle between the first and last points */
    selen = hypot(fv->endx-fv->startx,
                  fv->endy-fv->starty);
    fv->y[PF_SE_LEN] = selen;

    /* when the first and last points are very close,
    the angle features are muted so that they satisfy the stability criterion */
    factor = selen*selen / se_th_rolloff;
    if(factor > 1.0) factor = 1.0;
    factor = selen > EPS ? factor/selen : 0.0;
    fv->y[PF_SE_COS] = (fv->endx-fv->startx)*factor;
    fv->y[PF_SE_SIN] = (fv->endy-fv->starty)*factor;

    /* the remaining features have already been computed */
    fv->y[PF_LEN] = fv->path_r;
    fv->y[PF_TH] = fv->path_th;
    fv->y[PF_ATH] = fv->abs_th;
    fv->y[PF_SQTH] = fv->sharpness;

#ifdef PF_DUR
    fv->y[PF_DUR] = (fv->endtime-fv->starttime)*.01;
#endif

#ifdef PF_MAXV
    fv->y[PF_MAXV] = fv->maxv*10000;
#endif

    return fv->y;
}

File: symbolRecognizer2/fv.py
"""
Create a feature vector, useful for gesture classification,
from a sequence of points (e.g. mouse points).
"""
"""
/* indices into the feature Vector returned by FvCalc */

#define PF_INIT_COS	0 /* initial angle (cos) */
#define PF_INIT_SIN	1 /* initial angle (sin)  */
#define PF_BB_LEN	2 /* length of bounding box diagonal */
#define PF_BB_TH	3 /* angle of bounding box diagonal  */
#define PF_SE_LEN	4 /* length between start and end points */
#define PF_SE_COS	5 /* cos of angle between start and end points */
#define PF_SE_SIN	6 /* sin of angle between start and end points */
#define PF_LEN	7 /* arc length of path */
#define PF_TH	8 /* total angle traversed */
#define PF_ATH		9 /* sum of abs vals of angles traversed */
#define PF_SQTH	10 /* sum of squares of angles traversed */
"""
import math

PF_INIT_COS = 0
PF_INIT_SIN = 1
PF_BB_LEN = 2
PF_BB_TH = 3
PF_SE_LEN = 4
PF_SE_COS = 5
PF_SE_SIN = 6
PF_LEN = 7
PF_TH = 8
PF_ATH = 9
PF_SQTH = 10
PF_DUR = 11
PF_MAXV = 12

NFEATURES = 13
dist_sq_threshold = (3 * 3)
se_th_rolloff = (4 * 4)
EPS = math.pow(10, -4)


class FV:
    """
    Structure which holds intermediate results during feature vector calculation
    """
    def __init__(self):
        # The following are used in calculating the features
        self.startx = 0
        self.starty = 0
        self.starttime = 0

        # Initial angle to x axis, these are set after a few points and then left alone
        self.initial_sin = 0.0
        self.initial_cos = 0.0

        # These are updated incrementally upon every point
        self.npoints = 0
        # Differences: endx prevx, endy prevy
        self.dx2 = 0
        self.dy2 = 0
        self.magsq2 = self.dx2 * self.dx2 + self.dy2 * self.dy2
        # Last point added
        self.endx = 0
        self.endy = 0
        self.endtime = 0

        # Bounding box
        self.minx = 0
        self.maxx = 0
        self.miny = 0
        self.maxy = 0

        # Total length and rotation (in radians)
        self.path_r = 0
        self.path_th = 0
        # Sum of absolute values of path angles
        self.abs_th = 0
        # Sum of squares of path angles
        self.sharpness = 0
        # Maximum velocity
        self.maxv = 0

        # Actual feature vector
        self.y = [0.0 for i in range(NFEATURES)]

    def AddPoint(self, x, y, t):
        """
        Update a FV to reflect a new input point
        :param x: x-coordinate of the new point
        :param y: y-coordinate of the new point
        :param t: time values
        :return: void
        """
        self.npoints += 1

        # first point, initialize some vars
        if self.npoints == 1:
            self.starttime = t
            self.endtime = t
            self.startx = x
            self.endx = x
            self.minx = x
            self.maxx = x
            self.starty = y
            self.endy = y
            self.miny = y
            self.maxy = y
            return
        dx1 = x - self.endx
        dy1 = y - self.endy

        magsq1 = dx1 * dx1 + dy1 * dy1

        if magsq1 <= dist_sq_threshold:
            self.npoints -= 1
            return  # ignore a point close to the last point
        # Update some internal values if needed
        if x < self.minx:
            self.minx = x
        if x > self.maxx:
            self.maxx = x
        if y < self.miny:
            self.miny = y
        if y > self.maxy:
            self.maxy = y

        lasttime = self.endtime

        self.endtime = t

        d = math.sqrt(magsq1)
        self.path_r += d  # update path length feature

        # calculate initial theta when the third point is seen
        if self.npoints == 3:
            dx = x - self.startx
            dy = y - self.starty
            magsq = dx * dx + dy * dy
            if magsq > dist_sq_threshold:
                # find angle w.r.t.positive x axis e.g.(1, 0)
                recip = 1.0 / math.sqrt(magsq)
                self.initial_cos = dx * recip
                self.initial_sin = dy * recip

        if self.npoints >= 3:
            # / * update angle based features * /
            th = math.atan2(dx1 * self.dy2 - self.dx2 * dy1,
                            dx1 * self.dx2 + dy1 * self.dy2)
            absth = math.atan2(dx1 * self.dy2 - self.dx2 * dy1,
                               dx1 * self.dx2 + dy1 * self.dy2)
            if absth < 0:
                absth = -absth
            self.path_th += th
            self.abs_th += absth
            self.sharpness += th * th

            # Compute max velocity
            if self.endtime > lasttime and (d / (self.endtime - lasttime)) > self.maxv:
                self.maxv = d / (self.endtime - lasttime)

        # prepare for next iteration
        self.endx = x
        self.endy = y
        self.dx2 = dx1
        self.dy2 = dy1
        self.magsq2 = magsq1

        return

    def FvCalc(self):
        """
        Calculate and return a feature vector
        :return: feature vector
        """
        if self.npoints <= 1:
            return self.y   # A feature vector of all zeros
        self.y[PF_INIT_COS] = self.initial_cos
        self.y[PF_INIT_SIN] = self.initial_sin

        # Compute the length of the bounding box diagonal
        bblen = math.hypot(self.maxx - self.minx, self.maxy - self.miny)

        self.y[PF_BB_LEN] = bblen

        # The bounding box angle defaults to 0 for small gestures
        if bblen * bblen > dist_sq_threshold:
            self.y[PF_BB_TH] = math.atan2(self.maxy - self.miny, self.maxx - self.minx)

        # Compute the length and angle between the first and last points
        selen = math.hypot(self.endx - self.startx, self.endy - self.starty);
        self.y[PF_SE_LEN] = selen

        # When the first and last points are very close,
        # the angle features are muted so that they satisfy the stability criterion
        factor = selen * selen / se_th_rolloff
        if factor > 1.0:
            factor = 1.0

        if selen > EPS:
            factor /= selen
        else:
            factor = 0
        self.y[PF_SE_COS] = (self.endx - self.startx) * factor
        self.y[PF_SE_SIN] = (self.endy - self.starty) * factor

        # The remaining features have already been computed
        self.y[PF_LEN] = self.path_r
        self.y[PF_TH] = self.path_th
        self.y[PF_ATH] = self.abs_th
        self.y[PF_SQTH] = self.sharpness

        self.y[PF_DUR] = (self.endtime - self.starttime) * .01

        self.y[PF_MAXV] = self.maxv * 10000

        return self.y
